### PR TITLE
MULE-16321: Refactor how polices chain link to the `next` processor.

### PR DESCRIPTION
**Pull request by elrodro83**
* Add `clearVariables` on event builder.
* Remove `PolicyStateHandler` that introduced contention in favor of event's internalParameters to pass state around.
* Simplification of the polices code, some flat maps/child contexts removed.

Includes fix for MULE-16260

### CODE REVIEW

File: modules/deployment-model-impl/src/main/java/org/mule/runtime/module/deployment/impl/internal/policy/DefaultApplicationPolicyInstance.java [hunk 1]
```diff
@@ -163,7 +152,7 @@ private void enableNotificationListeners(List<NotificationListener> notification
   }
 
   private void initPolicyInstance() throws InitialisationException {
-    policyInstance = policyContext.getRegistry().lookupByType(DefaultPolicyInstance.class).get();
+    policyInstance = policyContext.getRegistry().lookupByType(PolicyInstance.class).get();
```
Comment (sonar-mulesoft) [inline modules/deployment-model-impl/src/main/java/org/mule/runtime/module/deployment/impl/internal/policy/DefaultApplicationPolicyInstance.java]: Bug Bug: Call "Optional#isPresent()" before accessing the value. ([squid:S3655](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3AS3655))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7629&issues=AWiGlnbUotIXnmsUonic&open=AWiGlnbUotIXnmsUonic)

[ISSUE_KEY]: AWiGlnbUotIXnmsUonic

File: core/src/main/java/org/mule/runtime/core/internal/policy/DefaultPolicyInstance.java [hunk 2]
```diff
@@ -214,9 +218,9 @@ public ProcessingStrategy getProcessingStrategy() {
     private static Logger LOGGER = getLogger(PolicyProcessingStrategy.class);
```
Comment (sonar-mulesoft) [inline core/src/main/java/org/mule/runtime/core/internal/policy/DefaultPolicyInstance.java]: Code Smell Code Smell: Rename this field "LOGGER" to match the regular expression '^[a-z][a-zA-Z0-9]*$'. ([squid:S3008](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3AS3008))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7629&issues=AWiGlnXCotIXnmsUoniU&open=AWiGlnXCotIXnmsUoniU)

[ISSUE_KEY]: AWiGlnXCotIXnmsUoniU

File: core/src/main/java/org/mule/runtime/core/internal/policy/DefaultOperationPolicyProcessorFactory.java [hunk 3]
```diff
@@ -18,19 +17,13 @@
...
-
   /**
    * Creates a new {@link Processor} from an operation {@link Policy}.
-   *
-   * @param policyStateHandler the state handler to use for keeping track of the policy chain event modifications.
    */
-  public DefaultOperationPolicyProcessorFactory(PolicyStateHandler policyStateHandler) {
-    this.policyStateHandler = policyStateHandler;
-  }
+  public DefaultOperationPolicyProcessorFactory() {}
```
Comment (sonar-mulesoft) [inline core/src/main/java/org/mule/runtime/core/internal/policy/DefaultOperationPolicyProcessorFactory.java]: Code Smell Code Smell: Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. ([squid:S1186](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3AS1186))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7629&issues=AWiGlnW3otIXnmsUoniS&open=AWiGlnW3otIXnmsUoniS)

[ISSUE_KEY]: AWiGlnW3otIXnmsUoniS

File: core/src/main/java/org/mule/runtime/core/api/policy/PolicyChain.java [hunk 4]
```diff
@@ -115,20 +117,20 @@ public CoreEvent process(CoreEvent event) throws MuleException {
...
   @Override
   public Publisher<CoreEvent> apply(Publisher<CoreEvent> publisher) {
-    return from(publisher)
+    return Flux.from(publisher)
+        .doOnNext(pushBeforeNextFlowStackElement()
+            .andThen(req -> ((BaseEventContext) req.getContext())
+                .onResponse((resp, t) -> popFlowFlowStackElement().accept(req)))
+            .andThen(notificationHelper.notification(PROCESS_START)))
+        // TODO MULE-16370 remove this flatMap
         .flatMap(event -> {
```
Comment (sonar-mulesoft) [inline core/src/main/java/org/mule/runtime/core/api/policy/PolicyChain.java]: Code Smell Code Smell: Remove useless curly braces around statement and then remove useless return keyword ([squid:S1602](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3AS1602))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7629&issues=AWiGlnVfotIXnmsUoniR&open=AWiGlnVfotIXnmsUoniR)

[ISSUE_KEY]: AWiGlnVfotIXnmsUoniR

File: core/src/main/java/org/mule/runtime/core/internal/policy/DefaultPolicyInstance.java [hunk 5]
```diff
@@ -77,18 +73,22 @@
...
 
+  private String name;
+
   private PolicyChain operationPolicyChain;
   private PolicyChain sourcePolicyChain;
 
-  private FlowConstructStatistics flowConstructStatistics = new DefaultFlowConstructStatistics("policy", getName());
-  private String name = "proxy-policy-" + UUID.getUUID();
+  private FlowConstructStatistics flowConstructStatistics;
   private MuleContext muleContext;
```
Comment (sonar-mulesoft) [inline core/src/main/java/org/mule/runtime/core/internal/policy/DefaultPolicyInstance.java]: Code Smell Code Smell: Remove this use of "MuleContext"; it is deprecated. ([squid:CallToDeprecatedMethod](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3ACallToDeprecatedMethod))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7629&issues=AWiGlnXCotIXnmsUoniX&open=AWiGlnXCotIXnmsUoniX)

[ISSUE_KEY]: AWiGlnXCotIXnmsUoniX

File: core/src/main/java/org/mule/runtime/core/internal/policy/DefaultPolicyInstance.java [hunk 6]
```diff
@@ -214,9 +218,9 @@ public ProcessingStrategy getProcessingStrategy() {
     private static Logger LOGGER = getLogger(PolicyProcessingStrategy.class);
     private static int SCHEDULER_BUSY_RETRY_INTERVAL_MS = 2;
```
Comment (sonar-mulesoft) [inline core/src/main/java/org/mule/runtime/core/internal/policy/DefaultPolicyInstance.java]: Code Smell Code Smell: Rename this field "SCHEDULER_BUSY_RETRY_INTERVAL_MS" to match the regular expression '^[a-z][a-zA-Z0-9]*$'. ([squid:S3008](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3AS3008))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7629&issues=AWiGlnXCotIXnmsUoniV&open=AWiGlnXCotIXnmsUoniV)

[ISSUE_KEY]: AWiGlnXCotIXnmsUoniV

File: core/src/main/java/org/mule/runtime/core/internal/policy/PolicyNextActionMessageProcessor.java [hunk 7]
```diff
@@ -94,54 +88,64 @@ private String toPolicyLocation(ComponentLocation componentLocation) {
...
+        .doOnNext(event -> {
           popBeforeNextFlowFlowStackElement().accept(event);
           notificationHelper.notification(BEFORE_NEXT).accept(event);
+        })
+
+        .flatMap(event -> subscriberContext().flatMap(ctx -> Mono.just(event).transform(ctx.get(POLICY_NEXT_OPERATION)))
+            .cast(CoreEvent.class)
+            .doOnError(MessagingException.class, this::handleExceptionInNext))
+
+        // TODO MULE-16371 Replace the flatMap above with this compose. Review how to handle the error from the flow in source
```
Comment (sonar-mulesoft) [inline core/src/main/java/org/mule/runtime/core/internal/policy/PolicyNextActionMessageProcessor.java]: Code Smell Code Smell: Complete the task associated to this TODO comment. ([squid:S1135](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3AS1135))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7629&issues=AWiGlnYDotIXnmsUonib&open=AWiGlnYDotIXnmsUonib)

[ISSUE_KEY]: AWiGlnYDotIXnmsUonib

File: core/src/main/java/org/mule/runtime/core/api/policy/PolicyChain.java [hunk 8]
```diff
@@ -115,20 +117,20 @@ public CoreEvent process(CoreEvent event) throws MuleException {
 
   @Override
   public Publisher<CoreEvent> apply(Publisher<CoreEvent> publisher) {
-    return from(publisher)
+    return Flux.from(publisher)
+        .doOnNext(pushBeforeNextFlowStackElement()
+            .andThen(req -> ((BaseEventContext) req.getContext())
+                .onResponse((resp, t) -> popFlowFlowStackElement().accept(req)))
+            .andThen(notificationHelper.notification(PROCESS_START)))
+        // TODO MULE-16370 remove this flatMap
```
Comment (sonar-mulesoft) [inline core/src/main/java/org/mule/runtime/core/api/policy/PolicyChain.java]: Code Smell Code Smell: Complete the task associated to this TODO comment. ([squid:S1135](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3AS1135))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7629&issues=AWiGlnVfotIXnmsUoniQ&open=AWiGlnVfotIXnmsUoniQ)

[ISSUE_KEY]: AWiGlnVfotIXnmsUoniQ

File: core/src/main/java/org/mule/runtime/core/internal/policy/DefaultSourcePolicyProcessorFactory.java [hunk 9]
```diff
@@ -17,19 +16,13 @@
...
-
   /**
    * Creates a new instance of the default factory for {@link SourcePolicy}s.
-   *
-   * @param policyStateHandler the state handler for policies.
    */
-  public DefaultSourcePolicyProcessorFactory(PolicyStateHandler policyStateHandler) {
-    this.policyStateHandler = policyStateHandler;
-  }
+  public DefaultSourcePolicyProcessorFactory() {}
```
Comment (sonar-mulesoft) [inline core/src/main/java/org/mule/runtime/core/internal/policy/DefaultSourcePolicyProcessorFactory.java]: Code Smell Code Smell: Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. ([squid:S1186](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3AS1186))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7629&issues=AWiGlnXlotIXnmsUonia&open=AWiGlnXlotIXnmsUonia)

[ISSUE_KEY]: AWiGlnXlotIXnmsUonia

**Comment by elrodro83** [comment]
--validate

**Pull request merged**